File: app/src/main/java/me/kirimin/mitsumine/ui/fragment/FeedFragment.java
package me.kirimin.mitsumine.ui.fragment;

import java.util.List;

import android.os.Bundle;

import me.kirimin.mitsumine.db.FeedDAO;
import me.kirimin.mitsumine.db.NGWordDAO;
import me.kirimin.mitsumine.model.Feed;
import me.kirimin.mitsumine.network.BookmarkFeedAccessor;
import me.kirimin.mitsumine.network.BookmarkFeedAccessor.CATEGORY;
import me.kirimin.mitsumine.network.BookmarkFeedAccessor.TYPE;
import me.kirimin.mitsumine.network.RequestQueueSingleton;
import me.kirimin.mitsumine.util.FeedFunc;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;

public class FeedFragment extends AbstractFeedFragment {

    public static FeedFragment newFragment(CATEGORY category, TYPE type) {
        FeedFragment fragment = new FeedFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(CATEGORY.class.getCanonicalName(), category);
        bundle.putSerializable(TYPE.class.getCanonicalName(), type);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    void requestFeed() {
        showRefreshing();
        CATEGORY category = (CATEGORY) getArguments().getSerializable(CATEGORY.class.getCanonicalName());
        TYPE type = (TYPE) getArguments().getSerializable(TYPE.class.getCanonicalName());
        final List<Feed> readFeedList = FeedDAO.findAll();
        final List<String> ngWordList = NGWordDAO.findAll();
        BookmarkFeedAccessor
                .requestCategory(RequestQueueSingleton.getRequestQueue(getActivity()), category, type)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMap(FeedFunc.mapToFeedList())
                .filter(FeedFunc.notContains(readFeedList))
                .filter(FeedFunc.notContainsWord(ngWordList))
                .toList()
                .subscribe(new Action1<List<Feed>>() {
                    @Override
                    public void call(List<Feed> feedList) {
                        clearFeed();
                        setFeed(feedList);
                        dismissRefreshing();
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        dismissRefreshing();
                    }
                });
    }

    @Override
    boolean isUseReadLater() {
        return true;
    }

    @Override
    boolean isUseRead() {
        return true;
    }
}

File: app/src/main/java/me/kirimin/mitsumine/ui/fragment/FeedFragment.kt
package me.kirimin.mitsumine.ui.fragment


import android.os.Bundle

import me.kirimin.mitsumine.db.FeedDAO
import me.kirimin.mitsumine.db.NGWordDAO
import me.kirimin.mitsumine.network.api.FeedApi
import me.kirimin.mitsumine.network.api.FeedApi.CATEGORY
import me.kirimin.mitsumine.network.api.FeedApi.TYPE
import me.kirimin.mitsumine.util.FeedUtil
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import java.io.Serializable

public class FeedFragment : AbstractFeedFragment() {

    var subscriptions = CompositeSubscription()

    override fun onDestroyView() {
        subscriptions.unsubscribe()
        super.onDestroyView()
    }

    override fun requestFeed() {
        showRefreshing()
        val category = arguments.getSerializable(CATEGORY::class.java.canonicalName) as CATEGORY
        val type = arguments.getSerializable(TYPE::class.java.canonicalName) as TYPE
        val readFeedList = FeedDAO.findAll()
        val ngWordList = NGWordDAO.findAll()
        subscriptions.add(FeedApi.requestCategory(category, type)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .filter { feed -> !FeedUtil.contains(feed, readFeedList) && !FeedUtil.containsWord(feed, ngWordList) }
                .toList()
                .subscribe ({ feedList ->
                    clearFeed()
                    setFeed(feedList)
                    dismissRefreshing()
                }, { e -> dismissRefreshing() })
        )
    }

    override fun isUseReadLater(): Boolean {
        return true
    }

    override fun isUseRead(): Boolean {
        return true
    }

    companion object {

        public fun newFragment(category: CATEGORY, type: TYPE): FeedFragment {
            val fragment = FeedFragment()
            val bundle = Bundle()
            bundle.putSerializable(CATEGORY::class.java.canonicalName, category as Serializable)
            bundle.putSerializable(TYPE::class.java.canonicalName, type as Serializable)
            fragment.arguments = bundle
            return fragment
        }
    }
}